### PR TITLE
Notification system

### PR DESCRIPTION
**Pull request by jbesraa**
full feature branch

### CODE REVIEW

File: api/controller/archive_notification.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+const Controller = require('../controller/controller');
+const { ArchiveNotification } = require('../model');
+
+class ArchiveNotificationController extends Controller {
```
Comment (dortheimer) [inline api/controller/archive_notification.js:4]: Add documentation, what does this controller do?

File: api/model/archive_notification.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+const Model = require('./base_model');
+const Exception = require('./exception');
+
+class ArchiveNotification extends Model {
```
Comment (dortheimer) [inline api/model/archive_notification.js:4]: Add documentation, what does this model represent?

File: api/controller/archive_notification.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+}
```
Comment (dortheimer) [inline api/controller/archive_notification.js:5]: A controller may also hold a creation method that is a factory for new models.

File: api/model/notification.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+const Model = require('./base_model');
+
+class Notification extends Model {
```
Comment (dortheimer) [inline api/model/notification.js:3]: Add documentation, what does this model represent?
What is the relationship with ArchiveNotification?

File: api/service/database.js [hunk 5]
```diff
@@ -1,14 +1,20 @@
 const Config = require('../lib/config');
-const Knex = require('knex')({
+
```
Comment (dortheimer) [inline api/service/database.js:2]: Why was this file changed?

**Comment by jbesraa** [comment]
@dortheimer this is a WIP branch and very far from completion, not ready yet for review

**Comment by dortheimer** [comment]
Sure, but I think that some of my comments could have value for you going on.

**Comment by jbesraa** [comment]
Sure thing. very welcome to write comments (:

File: api/service/database.js (hunk 5 above)
Comment (jbesraa) [reply api/service/database.js:2]: Mainly for readability

**Comment by jbesraa** [comment]
@dortheimer  here is the spec out of Notification feature: https://docs.google.com/document/d/1Y3a2hF48-8jCoN20y0eDtfQkkDnuwWHEbp4fdxYp4uI/edit?usp=sharing

I think it will answer some of your questions, feel free to comment if anything is not clear

**Comment by jbesraa** [comment]
@florpor  the problem with the crawler: `geom` = ST_GeomFromGeoJSON('{\"type\":\"Polygon\",\"coordinates\":[[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]]}')

the coordinates after parsing are empty. if you remove the guards(all line in line 38 and el.length > 0 in line 56) that I added in the parsing file(bookshelf-mysql-gis.js), you will get a different error.

**Comment by jbesraa** [comment]
@florpor this is good to go.
one thing we should check in order to make sure its working on production is the fact that we added new properties to some old tables(plan, person..),so maybe we need to do some migration. do we have staging env that we can test stuff on before we go to prod?

File: server/api/model/person.js [hunk 6]
```diff
@@ -11,17 +11,11 @@ const Exception = require('./exception');
...
-			// firstName: [
-			//   'required', 'string'kshelf
-			// ],
-			// lastName: [
-			//   'required', 'string'
-			// ],
 			email: ['required', 'email'],
 			password: ['required', 'string'],
-			status: ['required', 'integer'],
+			status: 'integer',
```
Comment (florpor) [inline server/api/model/person.js:18]: why was this change made? status is the user's activation status

File: server/api/model/plan.js [hunk 7]
```diff
@@ -73,12 +87,56 @@ class Plan extends Model {
...
+	handleUpdatedPlan (model) {
+		const planId = model.id;
+		const { users } = this.getUsersInPlanArea(model);
+		const types = this.getPlanUpdateTypes(model);
+		for(let type of types) {
+			Notification.createNotifications({ users, planId, type });
+		}
+	}
+
+	getUsersInPlanArea () {
```
Comment (florpor) [inline server/api/model/plan.js:114]: what is this for and why is the person_id hardcoded?

File: server/tests/tables_structs/alert_struct.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+const AlertStruct = function(table) {
+	table.increments();
+	table.integer('person_id');
+	table.string('address');
+	table.integer('radius');
+	table.binary('geom');
```
Comment (florpor) [inline server/tests/tables_structs/alert_struct.js:6]: maybe alerts aren't updated so the error we had with plan crawling never happens here, but should be changed to `table.specificType('geom', 'GEOMETRY');`

**Review by florpor — CHANGES_REQUESTED** [review]
see comments.
and also besides those - it seems the migration changes are not included here?

File: server/api/controller/plan.js [hunk 9]
```diff
@@ -64,9 +63,6 @@ class PlanController extends Controller {
 			type: 'Polygon',
 			coordinates: [points]
 		};
-		if (!GJV.valid(geojson)) {
```
[the remark this answers is not in the document]
Comment (jbesraa) [reply server/api/controller/plan.js:67]: I think I added that and when I tested it with invalid geojsons it didnt catch it
do you find it useful?

**Comment by jbesraa** [comment]
@florpor all of the above visited and fixed.

side note: Github actions are probably the worst CI I worked with. keep crashing for no reason. I strongly recommend to move to circleci/travis/.. something else
so sad that micr****t is ruining github ):

**Comment by jbesraa** [comment]
all green (:

File: server/api/controller/sign.js [hunk 10]
```diff
@@ -33,7 +33,7 @@ class SignController extends Controller {
 				// if there is a user but active, this will return an error
 
 				return this.model
-					.forge(req.body)
+					.forge({ ...req.body, status:0})
```
Comment (jbesraa) [inline server/api/controller/sign.js:36]: @florpor am not sure about this one
the status is required but its not part of the form the front end sends

File: server/api/controller/plan.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (florpor) [reply server/api/controller/plan.js:67]: i think you're confusing this with another test, as this one seems to have been [added in february](https://github.com/meirim-org/meirim/commit/b6f4f525d07b5860ca1df41e8185d972552a1b26#diff-3e9d9092c88f5d1e691049a150a6ef70e24feecb4a7d602bfaef31190d6925efR67).
i'm not sure what exact validations it runs or what was the problem it was introduced because of, but let's keep it.

File: server/api/controller/sign.js (hunk 10 above)
Comment (florpor) [reply server/api/controller/sign.js:36]: it's assigned in the `assignValues` function of the model class [here](https://github.com/meirim-org/meirim/blob/28e3320a1680bebf95f38ec28e30e20130e387e8/server/api/model/person.js#L45).
tests run fine without this change as far as i can see

File: .gitignore [hunk 11]
```diff
@@ -1,4 +1,5 @@
-node_modules
+server/node_modules
+client/node_modules
```
Comment (florpor) [inline .gitignore:2]: this is fine, but just know that the plain "node_modules" would ignore any file or folder named node_modules in the source tree:
`$ git check-ignore -v client/node_modules
.gitignore:1:node_modules	client/node_modules`

File: server/api/controller/sign.js (hunk 10 above)
Comment (jbesraa) [reply server/api/controller/sign.js:36]: the E2E tests? weird! they were failing for me before this change

**Comment by florpor** [comment]
@jbesraa great. see the two notes about geojson-validation and person status field, and also please add the new migration file(s?) to the pr

**Comment by jbesraa** [comment]
- [ ] geojson validation
- [ ] person status
- [ ] migrations

**Review by florpor — APPROVED** [review]

**Comment by florpor** [comment]
@jbesraa i see that the `static` modifier of person's rules function simply makes it not be checked and so it works.. ☉ ‿ ⚆
i'll fix that another time and reinstate the skipped test.

is this pr ready to be merged?

**Comment by jbesraa** [comment]
@florpor yea the static basically makes the rules function irrelevant. I think it also emphasis the problematic structure that the DB have.
atm we have 4 sources of 'truth' regarding the db: 1) the db it self 2) migration files 3) bookshelf setup 4) test structs. that can make things very tricky in the (near) future

yea PR is ready. should I merge it and u can test it on staging?
